source/s2n/s2n_tls_channel_handler.c

```diff
@@ -319,6 +320,11 @@ static void s_on_negotiation_result(
 static int s_drive_negotiation(struct aws_channel_handler *handler) {
     struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
 
+    if (s2n_handler->negotiation_failed || s2n_handler->negotiation_finished) {
+        AWS_LOGF_TRACE(AWS_LS_IO_TLS, "id=%p: Connection is not in a state to continue negotiation", (void *)handler);
```

Comment:
This will not happen often and if it does the log message isn't going to be your bottleneck. I wouldn't mind an assert here, but at a minimum i think this is ERROR or WARN

Reply:
also AWS_UNLIKELY would probably apply here as well?